### PR TITLE
WIP: Fixes #287 test time not accurate

### LINKED ISSUE #287: Test time is not accurate for first test
The first test always shows up as taking more time because of the overhead involved in loading the test file, running any pre-test code, etc.
### Repro

```
Start-Sleep -Seconds 2

Describe "Pester Test Time" {
    It "Is the First Test" {
        $true | Should be $true
    }

    It "Is the second Test" {
        $true | Should be $true
    }
}
```
### Comparison

```
Describe "Pester Test Time without Sleeping" {
    It "Is the First Test" {
        $true | Should be $true
    }

    It "Is the second Test" {
        $true | Should be $true
    }
}
```
### Results

```
Describing Pester Test Time
 [+] Is the First Test 2.06s
 [+] Is the second Test 6ms
Describing Pester Test Time without Sleeping
 [+] Is the First Test 63ms
 [+] Is the second Test 5ms
```

### PR DESCRIPTION
**Pull request by Tiberriver256**
## 1. General summary of the pull request

Fixes #287 - Test time is not accurate on the first run

I kept the concept of a single timer but split the timestamps into two places so that we get an accurate total test time as well as an accurate Test time.

**Test time** is evaluated based on timestamps in the EnterTest and LeaveTest functions.

**Total test time** is evaluated based on timestamps in the EnterTestGroup and LeaveTestGroup functions.

I also created two new tests for each case. Hopefully I got them in a section of the test that makes sense. Please let me know if there is anything I need to adjust.

### CODE REVIEW

File: Functions/PesterState.Tests.ps1 [hunk 1]
```diff
@@ -72,6 +72,36 @@ InModuleScope Pester {
         context "adding test result" {
             $p.EnterTestGroup('Describe', 'Describe')
 
+            it "times test accurately within 10 milliseconds" {
```
Comment (nohwnd) [inline Functions/PesterState.Tests.ps1]: The test placement is fine I think, but on a first look I am not sure what is being tested. Could you explain how that test works in a comment so we can refer to it when the test starts failing please?

File: Functions/PesterState.Tests.ps1 [hunk 2]
```diff
@@ -72,6 +72,36 @@ InModuleScope Pester {
...
+                $p.EnterTest()
+                $Time = Measure-Command -Expression {
+                    Start-Sleep -Milliseconds 100
+                }
+                $p.LeaveTest()
+                $p.AddTestResult("result","Passed",$null)
+                $result = $p.TestResult[-1]
+                [int]($result.time.TotalMilliseconds) | Should -BeGreaterOrEqual ($Time.Milliseconds-10)
```
Comment (nohwnd) [inline Functions/PesterState.Tests.ps1]: Why are the casts to `[int]` needed?

Comment (Tiberriver256) [reply Functions/PesterState.Tests.ps1]: I forgot to remove them from an earlier test method.

File: Functions/PesterState.Tests.ps1 (hunk 1 above)
Comment (Tiberriver256) [reply Functions/PesterState.Tests.ps1]: Definitely!

I hope you don't think I went overboard on the comments there but I added quite a few. I also discovered a misconception I had on the group level testing which contributed to it being so far off (+- 200 milliseconds) and now the test runs at +-10 milliseconds for the group as well as the individual test where it should be.

Let me know if any of my thinking for how the timings work is off but I think it's looking pretty stinking accurate on the timings now.

Thanks!

**Comment by Tiberriver256** [comment]
Actually.... One more thing I just noticed on the timings from an nunit xml. The the TestGroup level timings are the sum of the tests within that TestGroup. Is that the way it should be? or should the TestGroup time include the time taken for BeforeAll / BeforeEach blocks?

I'll give a couple of examples down below. Just let me know if it should look like example 1 or example 2 and I'll drop in a quick fix or leave it as is.

## Test

```pwsh
Describe "Test" {
        BeforeAll { Start-Sleep -Seconds 1  } 
        
        Context "Sleeping" {
           It "Sleeps for a second" { Start-Sleep -Seconds 1 }
        }        
}
```

## Example 1

* Timing at top level reflects total time for all tests including BeforeAll of 2 seconds
* Timing at script level test group **.\dummytest.ps1** reflects total time for all tests including BeforeAll of 2 seconds
* Timing at describe level test group **Test** reflects total time for all tests including BeforeAll of 2 seconds
* Timing at context level test group **Sleeping** reflects total time for all nested scriptblocks of 1 second
* Timing at test **"Sleeps for a second** reflects total time of the single test of 1 second

```xml
. . .
  <test-suite type="TestFixture" name="Pester" executed="True" result="Success" success="True" time="2" asserts="0" description="Pester">
    <results>
      <test-suite type="TestFixture" name=".\dummytest.ps1" executed="True" result="Success" success="True" time="2" asserts="0" description=".\dummytest.ps1">
        <results>
          <test-suite type="TestFixture" name="Test" executed="True" result="Success" success="True" time="2" asserts="0" description="Test">
            <results>
              <test-suite type="TestFixture" name="Sleeping" executed="True" result="Success" success="True" time="1" asserts="0" description="Sleeping">
                <results>
                  <test-case description="Sleeps for a second" name="Test.Sleeping.Sleeps for a second" time="1" asserts="0" success="True" result="Success" executed="True" />
                </results>
              </test-suite>
            </results>
          </test-suite>
        </results>
      </test-suite>
    </results>
  </test-suite>
. . .
```

## Example 2

* Timing at top level reflects total time for all tests including BeforeAll of 2 seconds
* Timing at script level test group **.\dummytest.ps1** reflects total time for all tests of 1 second
* Timing at describe level test group **Test** reflectstotal time for all tests of 1 second
* Timing at context level test group **Sleeping** reflects total time for all nested tests of 1 second
* Timing at test **"Sleeps for a second** reflects total time of the single test of 1 second

```xml
. . .
  <test-suite type="TestFixture" name="Pester" executed="True" result="Success" success="True" time="2" asserts="0" description="Pester">
    <results>
      <test-suite type="TestFixture" name=".\dummytest.ps1" executed="True" result="Success" success="True" time="1" asserts="0" description=".\dummytest.ps1">
        <results>
          <test-suite type="TestFixture" name="Test" executed="True" result="Success" success="True" time="1" asserts="0" description="Test">
            <results>
              <test-suite type="TestFixture" name="Sleeping" executed="True" result="Success" success="True" time="1" asserts="0" description="Sleeping">
                <results>
                  <test-case description="Sleeps for a second" name="Test.Sleeping.Sleeps for a second" time="1" asserts="0" success="True" result="Success" executed="True" />
                </results>
              </test-suite>
            </results>
          </test-suite>
        </results>
      </test-suite>
    </results>
  </test-suite>
. . .
```

I'm pretty sure example 1 is what we're going for here but wanted to double check.

**Comment by nohwnd** [comment]
For me the time should be all the time spent in user code, because the whole test does take that long to run. On the other hand some people do analysis on test runs and try to find outliers that take unnecessary amount of time and including setup in the test time makes it difficult to determine what is happening, whether the test is slower than the others or just does expensive setup. At the moment I am working on new internals for Describe and It, and there it should be easier to track how long each user step (be it setup, test or teardown took) and how much time the framework took, analysis can then be done on the result of the test run rather than the nUnit output, the nUnit output should imho report the setup as part of the test.

**Comment by nohwnd** [comment]
@Tiberriver256 I think it should look like example 1, are changes needed?

**Comment by nohwnd** [comment]
@Tiberriver256 Sorry to ping you again, but I'd like to release soon, and this would be nice to include. Are there any changes needed to make it work like example 1? Thx

**Comment by Tiberriver256** [comment]
Hey @nohwnd, sorry for the eternal delay. Yes, there are some minor changes. I'll get them up here tonight.

**Comment by Tiberriver256** [comment]
Wow, that took a lot more tinkering to get the tests passing again after merging with the latest version on master. Let me know how it looks!

**Comment by nohwnd** [comment]
Thank you, merged. 🎉

**Pull request merged**